lib/Utils/AttributeUtils.cpp

```diff
@@ -0,0 +1,116 @@
+#include "llvm/include/llvm/ADT/TypeSwitch.h"         // from @llvm-project
+#include "llvm/include/llvm/Support/LogicalResult.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Affine/IR/AffineOps.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Attributes.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Block.h"       // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"       // from @llvm-project
+#include "mlir/include/mlir/Interfaces/FunctionInterfaces.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+int getOperandNumber(Operation *op, Value value) {
+  for (OpOperand &operand : op->getOpOperands()) {
+    if (operand.get() == value) return operand.getOperandNumber();
+  }
+  return -1;
+}
+
+int getResultNumber(Operation *op, Value value) {
+  for (auto result : op->getOpResults()) {
+    if (result == value) return result.getResultNumber();
+  }
+  return -1;
+}
+
+Attribute findAttributeForBlockArgument(BlockArgument blockArg,
+                                        StringRef attrName) {
+  auto *parentOp = blockArg.getOwner()->getParentOp();
+  assert(parentOp != nullptr &&
+         "Missing parent op! Was this value not properly remapped?");
+
+  return llvm::TypeSwitch<Operation *, Attribute>(parentOp)
+      .Case<FunctionOpInterface>([&](auto op) -> Attribute {
+        return op.getArgAttr(blockArg.getArgNumber(), attrName);
+      })
+      // AffineForOp needs a special case because its operands do not
+      // line up perfectly with its block arguments.
+      // This could be potentially improved by adding an interface method to
+      // OperandAndResultAttrInterface to map from block arguments to operands
+      .Case<affine::AffineForOp>([&](affine::AffineForOp op) -> Attribute {
+        // For op has as its first block argument the induction
+        // variable, which does not correspond to a single operand.
+        if (blockArg.getArgNumber() == 0) return nullptr;
+
+        auto argAttrInterface = cast<OperandAndResultAttrInterface>(*op);
+        auto initArg = op.getInits()[blockArg.getArgNumber() - 1];
+        int operandNumber = getOperandNumber(op, initArg);
+        if (operandNumber == -1) return nullptr;
+        return argAttrInterface.getOperandAttr(operandNumber, attrName);
+      })
+      .Case<OperandAndResultAttrInterface>([&](auto op) -> Attribute {
+        return op.getOperandAttr(blockArg.getArgNumber(), attrName);
+      })
+      .Default([](Operation *op) { return nullptr; });
+}
+
+Attribute findAttributeForResult(Value result, StringRef attrName) {
+  auto *parentOp = result.getDefiningOp();
+  assert(parentOp &&
+         "Missing defining op, but the input value is not a BlockArgument. "
+         "This is madness.");
+
+  return llvm::TypeSwitch<Operation *, Attribute>(parentOp)
+      .Case<OperandAndResultAttrInterface>([&](auto op) -> Attribute {
+        int resultNumber = getResultNumber(op, result);
```

Comment:
this is so weird but it seems in the codebase that you can just cast a value as an OpResult and it will do this for you automatically when you call `getResultNumber`. So once you debug assert about the defining op, you can just cast the result as an opresult.

e.g. https://github.com/llvm/llvm-project/blob/718c4ed8a0d69948ab4b53e3c8666fd72613b7f7/mlir/lib/Dialect/Bufferization/IR/BufferizableOpInterface.cpp#L154

Reply:
🤯🤯🤯🤯🤯

Reply:
Fixed in 7931d365